### PR TITLE
refactor: upgrade ts-graphql-plugin to get TypedQueryDocumentNode type aliases

### PR DESCRIPTION
**Pull request by hallettj**
This change reduces the necessary type assertion on each GraphQL query from three lines to one, and provides the same type inference.

### CODE REVIEW

File: template/packages/frontend/typings/typed-document-node.d.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
+/*
+ * Ok, are you ready for this? The short version is that we should be able to
```
Comment (xasopheno) [inline template/packages/frontend/typings/typed-document-node.d.ts:2]: Uh oh...I hope so...

File: template/packages/frontend/typings/typed-document-node.d.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+ * added in `graphql` v15.4.0 - and libraries haven't yet updated to consume it
+ * so they don't benefit from the extra type information. (An added wrinkle is
+ * that importing this type introduces a minimum version constraint on `graphql`
+ * which will discourage library authors from using the type for a while.)
+ *
+ * But there is hope! `TypedQueryDocumentNode` is based on a type called
+ * [`TypedDocumentNode`][] from a third-party package,
+ * `@graphql-typed-document-node/core`. That package was made by the GraphQL
+ * Code Generator team as a way to get a common type in use until GraphQL adopted
+ * their own version. (Did you know that Jesse was the one to originally suggest
```
Comment (xasopheno) [inline template/packages/frontend/typings/typed-document-node.d.ts:51]: Jesse: Resident Superhero

**Comment by xasopheno** [comment]
LGTM! @hallettj  Thanks for fighting the good fight. Excited to see this land. 

@mjhm @allonsy @matthandlersux @hlian @iphipps
This is a very fun PR. :+1:

**Comment by matthandlersux** [comment]
@hallettj this looks interesting. i saw the very long explanation as to what magic you get from this PR, but perhaps you could boil it down a bit more for me? my team uses `graphql-codegen`, would this be an improvement on that?

**Comment by hallettj** [comment]
@matthandlersux There are two reasons why I selected [ts-graphql-plugin](https://github.com/Quramy/ts-graphql-plugin/blob/master/README.md). The most important is that it emits types that lead to easy-to-understand error messages. As much as I like graphql-codegen, it composes types such that when you make a mistake with result data or with query variables the resulting error is difficult to understand. One of the pieces of feedback I got from the other devs on Nuloan was that error messages were a problem. OTOH ts-graphql-plugin generates simple interfaces for each query.

Another piece of feedback was that - not being familiar with GraphQL - the other devs were confused about what happens where, what each file is responsible for, and which things are happening on the server vs the client. To address that, and to hopefully speed of the development iteration cycle as much as possible, I've been aiming to reduce the number of files devs have to switch between to implement and to query the API. On the server side that means using a code-first strategy with [NestJS](https://docs.nestjs.com/graphql/quick-start) where resolver code defines both the API implementation, and the schema. On the client side I wanted to allow devs to write queries in TypeScript co-located with the components that use the queries if they so choose.

I also like the interactivity of ts-graphql-plugin. It provides two pieces: it generates types for queries, and it provides a TypeScript compiler plugin that checks queries as you type them, and shows field types on hover. (It also generates a report of all queries used in the app which might come in handy.)

There are some downsides: compiler plugins can't change the output of `tsc` so to catch potential problems in CI we also need an eslint plugin to do the same checks. There is also the matter where I had to write a [custom eslint plugin / autofix](https://github.com/Originate/eslint-plugin-ts-graphql) to automatically insert type assertions into query expressions; and I had to submit a PR to ts-graphql-plugin to add support for `TypedQueryDocumentNode` to avoid the need for manual intervention for hooking up types. We're trying to use off-the-shelf libraries as much as possible for stability, and to try to make sure all the pieces are documented and will be understandable to new engineers. My hope is that the custom eslint plugin, and the type definition file in this PR are small enough pieces of the overall puzzle that things will work out.

Another quite interesting library I looked at is [ts-gql](https://github.com/Thinkmill/ts-gql/blob/master/README.md). Like ts-graphql-plugin it supports writing queries inline in TypeScript code. It even has its own eslint autofix which is what inspired / provided the initial code for the plugin that I wrote. Instead of using a compiler plugin ts-gql uses eslint to provide feedback on query correctness - so you don't need a separate eslint plugin for that checking - but that does make it a little less interactive than ts-graphql-plugin. Ts-gql uses graphql-codegen to generate types with an extra piece similar to `TypedQueryDocumentNode`, and provides wrappers for GraphQL libraries that consume the types it generates to propagate result data and query variable types. Ts-gql would have been substantially less work to set up. But like I said, the error messages with ts-graphql-plugin are much easier to understand, and I think that is quite valuable.

There's actually one more library on my list that I found out about the other day that I've been meaning to evaluate: [typed-graphql-client](https://www.npmjs.com/package/graphql-typed-client). It generates a TypeScript DSL for a given schema which avoids a need for a compiler plugin or eslint checking which could potentially simplify things.

**Comment by matthandlersux** [comment]
thanks @hallettj, very comprehensive.

**Pull request merged**